### PR TITLE
Adding Configurable to set number of threads for ONNX

### PR DESCRIPTION
**Pull request by ChSonnabend**
It was noticed that upon running within singularity containers on a SLURM cluster, ONNX has difficulties to allocate the correct resources if the number of CPU's is limited. A manually set value for the number of available threads fixed this issue. This should not affect the code any further, since the default value lets ONNX choose the number of threads automatically.

### CODE REVIEW

**Comment by jgrosseo** [comment]
Hi Christian. How do we avoid that this is misused on the Grid?

**Comment by ChSonnabend** [comment]
> Hi Christian. How do we avoid that this is misused on the Grid?

Hi Jan-Fiete,
That's a very good point indeed. I haven't taken that into consideration yet. I think we could add a flag "runOnHyperloop" which could disable the usage of the number of threads automatically (just another condition in the if-statement in the header .cxx file). What do you think about this?
Potentially we wouldn't even need to use this feature at all on hyperloop, I only noticed this behaviour on the GSI SLURM cluster.
Are there maybe already some variables automatically set when an analysis is run on hyperloop? Then I could maybe use these...

**Comment by jgrosseo** [comment]
There are some environment variables present in a Grid job which one could use but at the moment we do not provide them on Hyperloop (could be done of course) so that it looks the same.

I flagged this because we had issues with ONNX Grid jobs at some point because they were pinning to the wrong CPU and then the threads were disabled. But I do not remember the details @saganatt, @mpuccio do you?

**Comment by jgrosseo** [comment]
If we could use the inverse trick for you, we may not even need a configurable...

**Comment by jgrosseo** [comment]
Discussing with Costin, on the Grid there is the variable `ALIEN_JDL_CPUCORES` which should be respected. We can also set it on Hyperloop. I guess if that variable is set, you should ignore the setting of the number of threads. What do you think?

**Comment by ChSonnabend** [comment]
> Discussing with Costin, on the Grid there is the variable `ALIEN_JDL_CPUCORES` which should be respected. We can also set it on Hyperloop. I guess if that variable is set, you should ignore the setting of the number of threads. What do you think?

Indeed, I think this is a good idea. I also feel much better if there is an automatic switch like this rather than the user having to specify it themselves. Just to be sure, this variable is only set on hyperloop right?
I will look up the variable and adjust the code.
Thank you for your help!

**Comment by jgrosseo** [comment]
> > Discussing with Costin, on the Grid there is the variable `ALIEN_JDL_CPUCORES` which should be respected. We can also set it on Hyperloop. I guess if that variable is set, you should ignore the setting of the number of threads. What do you think?
> 
> Indeed, I think this is a good idea. I also feel much better if there is an automatic switch like this rather than the user having to specify it themselves. Just to be sure, this variable is only set on hyperloop right? I will look up the variable and adjust the code. Thank you for your help!

It will be set in the Hyperloop test and for all Grid jobs, also those which go to GSI. Your observation above was I guess with locally submitting to the cluster and not through the Grid, right? In case of your local submissions, the variable is of course not set.

I propose that if that variable is set, you ignore the user configurable and if the user configurable is not the default value print a LOGP(warning) to inform the user that the variable is ignored.

What do you think?

File: Common/TableProducer/PID/pidTPC.cxx [hunk 1]
```diff
@@ -142,9 +148,25 @@ struct tpcPid {
...
+
     if (!useNetworkCorrection) {
       return;
     } else {
+
+      /// Testing hyperloop core settings
+      const char* alien_cores = gSystem->Getenv("ALIEN_JDL_CPUCORES");
+      if (alien_cores != NULL) {
+        std::string msg_cores = "Hyperloop-run detected! Number of cores = ";
+        LOGP(info, msg_cores + std::string(alien_cores));
```
Comment (jgrosseo) [inline Common/TableProducer/PID/pidTPC.cxx:161]:
```suggestion
        LOGP(info, "Hyperloop test/Grid job detected! Number of cores = {}. Setting threads anyway to 1.", alien_cores);
```

File: Common/TableProducer/PID/pidTPCFull.cxx [hunk 2]
```diff
@@ -142,9 +148,25 @@ struct tpcPidFull {
...
+
     if (!useNetworkCorrection) {
       return;
     } else {
+
+      /// Testing hyperloop core settings
+      const char* alien_cores = gSystem->Getenv("ALIEN_JDL_CPUCORES");
+      if (alien_cores != NULL) {
+        std::string msg_cores = "Hyperloop-run detected! Number of cores = ";
+        LOGP(info, msg_cores + std::string(alien_cores));
```
Comment (jgrosseo) [inline Common/TableProducer/PID/pidTPCFull.cxx:161]: Same as above if you like it ;)

Comment (ChSonnabend) [reply Common/TableProducer/PID/pidTPCFull.cxx:161]: Yes, thank you, will add that ;-)

[the remark this answers is not in the document]
Comment (jgrosseo) [reply Common/TableProducer/PID/pidTPCFull.cxx:161]: Makes it easier ;)

**Review by jgrosseo — APPROVED** [review]

**Comment by fcatalan92** [comment]
> I flagged this because we had issues with ONNX Grid jobs at some point because they were pinning to the wrong CPU and then the (additional?) threads were disabled. But I do not remember the details @saganatt, @mpuccio do you?

@jgrosseo The issue was for the treelite library and its interaction with jobs running on singularity containers. We do not use this library anymore in O2/O2Physics.
However, also ONNXRuntime has its own thread pool and similar issues could arise. IMHO, a central interface to ONNXRuntime is needed to enforce optimisation and implement eventual fixes in a single place.

**Comment by fgrosa** [comment]
> > I flagged this because we had issues with ONNX Grid jobs at some point because they were pinning to the wrong CPU and then the (additional?) threads were disabled. But I do not remember the details @saganatt, @mpuccio do you?
> 
> @jgrosseo The issue was for the treelite library and its interaction with jobs running on singularity containers. We do not use this library anymore in O2/O2Physics. However, also ONNXRuntime has its own thread pool and similar issues could arise. IMHO, a central interface to ONNXRuntime is needed to enforce optimisation and implement eventual fixes in a single place.

Hi @fcatalan92 @jgrosseo indeed the issue was spotted for the HFFilters (at least, might be not only). The CPU time was 600% without adding
```
https://github.com/AliceO2Group/O2Physics/blob/a8b591be5a0498dd1b443c40242261d88582ca98/EventFiltering/PWGHF/HFFilter.cxx#L234-L235
```
while it is close to 100% after setting them to 1.

**Review by jgrosseo — APPROVED** [review]

**Comment by jgrosseo** [comment]
Thanks for your replies. @ChSonnabend and me agreed to have the threads fixed to 1 for now on the Grid and we will discuss the strategy at a WP14 meeting.

**Pull request merged**